### PR TITLE
REF: Dispatch string methods to ExtensionArray

### LINKED ISSUE #36216: Refactor StringMethods for extension arrays
This is an issue to track a refactor of StringMethods that'll be necessary for ArrowStringArray to use pyarrow compute algorithms.

We'll need to update StringMethods

1. Extract the array from the Series / Index
2. Dispatch the string methods
3. Wrap the result

So in my proposal, we'll have something like

```python
class StringMethods:
    def __init__(self, data, ...):
        self._array = data.array
        ...

    def upper(self):
        return self._wrap_result(self._array._str.upper())
```

cc @xhochy. I have a branch started and will hopefully finish it off this week.

### PR DESCRIPTION
**Pull request by TomAugspurger**
This is a large refactor of our string methods. The goal is to have the `Series.str` accessor dispatch the actual compute down to the ExtensionArray. The motivation is to allow a Series backed by Arrow memory to use the Arrow string kernel (e.g. `Series.str.lower()` on an Arrow-backed StringArray should eventually call `pyarrow.compute.utf8_lower()`.)

To facilitate this, I made the following changes

1. Split `core/strings.py` into a sub package:
  - `core/strings/accessor.py`: Implements the `Series.str` accessor, which (mostly) just delegates to the EA and wraps
  - `core/strings/object_array.py`: Implements the string methods for object-type ndarray. 
  - `core/strings/categorical.py`, `core/strings/string_array.py`, implements categorical & StringArray-specific methods
2. Defines a new `ExtensionArray._str` extension point. This is where EAs get to take over and use their compute

Note that there are a few methods like `cat`, `extract`, and `extractall` that don't yet dispatch. I need to look into them a bit more, there implementation is a bit confusing.

Closes https://github.com/pandas-dev/pandas/issues/36216

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Making sure I understand: core.strings.accessor is mostly a cut/paste of the existing core.strings?  strings.object_array corresponds to the ndarray[object]-backed StringArray we use now?

Is strings.categorical for a Categorical backed by a StringArray or a StringArray backed by a Categorical?  Something else?

Is the accessor central to what you're doing here, or could in principle be separate?  I expect to have an opinion about this, but am not there yet.

**Comment by TomAugspurger** [comment]
> core.strings.accessor is mostly a cut/paste of the existing core.strings

Not quite. Anywhere you see things like 

```python
def method(self, ...):
    result = self._array._str.method(...)
    return self._wrap_result(result)
```

that `result = self._array._str.method` was previously something like `map(self._array, func)`, or something like `return str_method`.

The core methods `map_object` and `map_na` which were previously in `strings.py` were moved to `ObjectArrayMethods._map` and `StringArrayMethods._map`.

> Is strings.categorical for a Categorical backed by a StringArray or a StringArray backed by a Categorical? Something else?

It's for a supporting `Series[Categorical].str` and `CategoricalIndex.str`. Which I think is just a Categorical containing strings (object-dtype index for the categories).

> Is the accessor central to what you're doing here, or could in principle be separate?

I assume you mean the `_str` accessor on ExtensionArrays, since that's what's new? It's central in the sense that it and inheritance is the mechanism I've used to dispatch. We could do it other ways.

File: pandas/core/strings/object_array.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+
+    def get_dummies(self, sep="|"):
+        from pandas import Series
+
+        arr = Series(self._array).fillna("")
+        assert isinstance(arr, Series)  # fillna returns Optional[Series]
+        try:
+            arr = sep + arr + sep
+        except TypeError:
+            arr = sep + arr.astype(str) + sep
```
Comment (TomAugspurger) [inline pandas/core/strings/object_array.py:348]: @simonjayhawkins any idea why mypy is failing here?

```
pandas/core/strings/object_array.py:348: error: Item "None" of "Optional[Series]" has no attribute "astype"  [union-attr]
```

I would have thought the `assert isinstance(arr, Series)` would make this pass. Declaring `arr: Series = ...` doesn't fix it either.

Comment (simonjayhawkins) [reply pandas/core/strings/object_array.py:348]: Not sure why. but sep is Any, and `sep + arr + sep` is resolving to Any and arr is reset back to the original type for arr.

`arr = cast(Series, sep + arr + sep)` fixes till we have the binops typed and we can track the cast.

you could also do this instead of the earlier assert so that once Literal is used and the return type of fillna is correct for inplace, we can track the cast.

**Comment by jbrockmendel** [comment]
> I assume you mean the _str accessor on ExtensionArrays, since that's what's new? It's central in the sense that it and inheritance is the mechanism I've used to dispatch. We could do it other ways.

In the past we've been very conservative about adding things to EA.  I'd find this easier to think about if that could be considered orthogonally.

**Comment by TomAugspurger** [comment]
> In the past we've been very conservative about adding things to EA. I'd find this easier to think about if that could be considered orthogonally.

Just to clarify, I wouldn't consider this isn't part of the official EA interface (I'm not adding it to the docs, for example). If someone comes along with another EA for storing string arrays and want access to the pandas' `.str` namespace, then I'm happy to discuss adding it to the interface. Right now, we just have the 3-4ish internal users in pandas (object-dtype, string dtype, categorical, and soon ArrowStringArray).

We just need some way to translate `Series.str.<method>` to "call the right <method> for a Series backed by this extension array".

File: pandas/core/strings/base.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+from pandas._typing import Scalar
+
+from pandas.core.arrays.base import ExtensionArray
+
+
+class BaseStringArrayMethods(abc.ABC):
+    """
+    Base class for array _str accessor.
+
+    This is where ExtensionArrays can override the implementation of
```
Comment (jorisvandenbossche) [inline pandas/core/strings/base.py:15]:
```suggestion
    This is where our ExtensionArrays can override the implementation of
```

(that's maybe clearer that it's internal (for now) ..)

File: pandas/core/strings/object_array.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+        return self._map(lambda x: x.strip(to_strip))
+
+    def lstrip(self, to_strip=None):
+        return self._map(lambda x: x.lstrip(to_strip))
+
+    def rstrip(self, to_strip=None):
+        return self._map(lambda x: x.rstrip(to_strip))
+
+
+class ObjectProxy(PandasArray):
```
Comment (jorisvandenbossche) [inline pandas/core/strings/object_array.py:422]: Why does it need to subclass from PandasArray? Isn't only the `._str` attribute ever accessed?

Comment (jorisvandenbossche) [reply pandas/core/strings/object_array.py:422]: Ah, OK, the `self._array` in the ...ArrayMethods class is this object, right? Will comment on a line above then

File: pandas/core/strings/object_array.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+from pandas.core.accessor import CachedAccessor
+from pandas.core.arrays.numpy_ import PandasArray
+from pandas.core.strings.base import BaseStringArrayMethods
+
+
+class ObjectArrayMethods(BaseStringArrayMethods):
+    _default_na_value = np.nan
+
+    def _map(self, f, na_value=None, dtype=None):
+        arr = self._array  # object-dtype ndarray.
```
Comment (jorisvandenbossche) [inline pandas/core/strings/object_array.py:26]: this is actually an object-dtype ndarray *wrapped* in ObjectProxy/PandasArray?

File: pandas/core/strings/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+from .accessor import StringMethods
+from .base import BaseStringArrayMethods
+
+__all__ = ["StringMethods", "BaseStringArrayMethods"]
```
Comment (jorisvandenbossche) [inline pandas/core/strings/__init__.py:4]: Import here also the other classes that are used outside of pandas.core.strings (`StringArrayMethods`, `CategoricalStringMethods`) ?

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good to me! This will nicely allow to start overriding a subset of the methods with arrow implementations where available for the ArrowStringArray.

I suppose one possible alternative to the "proxy" class holding the methods that is added as `_str` on the ExtensionArrays would be to add those methods *directly* on the EA. That's one layer of indirection less, but on the other hand also adds a whole bunch of methods to those classes of course (but it's similar to eg DatetimeArray holding the .dt methods/attributes, and the Methods classes could be turned into a mixin)

File: pandas/core/strings/object_array.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/strings/object_array.py:26]: I'll need to check... My types say it should be an `ExtensionArray` (ObjectProxy is an ExtensionArray subclass). I think we have

```python
Series.str : pandas.core.strings.StringMethods
pandas.core.strings.StringMethods._array : ExtensionArray (specifically, ObjectProxy, Categorical, StringArray)
pandas.core.strings.StringMethods._array._str : BaseStringArrayMethods
pandas.core.strings.StringMethods._array._str._array : ExtensionArray
```

My head is hurting a bit from all the nesting. Surely this can be simplified?

Comment (jorisvandenbossche) [reply pandas/core/strings/object_array.py:26]: > My head is hurting a bit from all the nesting. Surely this can be simplified?

Well, with what I said above: define the methods directly on the ExtensionArray, so you would only have `pandas.core.strings.StringMethods._array.<method>` (without the extra `._str`). But of course that has other downsides.

**Comment by TomAugspurger** [comment]
So I think our two options are either

1. Attach an accessor to string EAs

```python
class BaseStringArrayMethods(abc.ABC):
     def __init__(self, array: ExtensionArray):
        self._array = array

class ObjectProxy(ExtensionArray):
   _str = CachedAccessor("str", ObjectArrayMethods)
```

2. A mixin, where we prefex each method with some string (to disambiguate from regular methods of the same name).

```python
class StringMethodsMixin:
    def _str_isnumber(self):
        ...

    def _str_repeat(self, n):
        ...

class StringArray(ExtensionArray, StringMethodsMixin):
    ...
```

Sounds like people are roughly prefer the second version?

**Review by jreback — CHANGES_REQUESTED** [review]
can you do a precursor PR which just moves things. I have no idea what really changed here.

**Comment by TomAugspurger** [comment]
Not easily, since it's more of a split than a move. Right now the `pandas/core/strings.py` contains

1. The Series.str accessor (the class StringMethods and its function signatures)
2. The *implementations* of these methods on arrays (the bodies of StringMethod's methods, all the top-level `str_<method>` methods, `na_map`, `map_object`, `map_na`.

This PR moves the `StringMethods` accessor to `pandas/core/strings/accessor.py` and the implementations to `pandas/core/strings/{base,object_array,string_array,categorical}.py`

File: pandas/core/strings/__init__.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/strings/__init__.py:4]: The others are gone now that we're using a methods mixing & inheritance.

**Review by TomAugspurger — COMMENTED** [review]
Updated to use a mixin (`ObjectStringArrayMixin`) that implements the pandas string methods on object-dtype ndarrays. EAs implementing the string methods (StringArray, Categorical, and PandasArray for object-dtype strings) inherit from that.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jorisvandenbossche — COMMENTED** [review]
The updates to use a mixin look good to me!

Can you update the docstring of BaseStringArrayMethods to reflect the new implementation?

File: pandas/core/strings/accessor.py [hunk 6]
```diff
@@ -0,0 +1,3078 @@
...
+
+
+def _get_single_group_name(rx):
+    try:
+        return list(rx.groupindex.keys()).pop()
+    except IndexError:
+        return None
+
+
+def _str_extract_noexpand(arr, pat, flags=0):
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/strings/accessor.py:2964]: See @TomAugspurger note in the top post: 

> Note that there are a few methods like cat, extract, and extractall that don't yet dispatch. I need to look into them a bit more, there implementation is a bit confusing.

File: pandas/core/arrays/numpy_.py [hunk 7]
```diff
@@ -398,6 +402,10 @@ def arithmetic_method(self, other):
 
     _create_comparison_method = _create_arithmetic_method
 
+    # ------------------------------------------------------------------------
+    # String methods interface
+    _str_na_value = np.nan
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py:407]: not pd.NA?

File: pandas/core/strings/base.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+import numpy as np
+
+from pandas._typing import Scalar
+
+
+class BaseStringArrayMethods(abc.ABC):
+    """
+    Base class for extension arrays implementing string methods.
+
+    This is our ExtensionArrays can override the implementation of
```
Comment (jbrockmendel) [inline pandas/core/strings/base.py:13]: "our" -> "so"?

File: pandas/core/arrays/numpy_.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/numpy_.py:407]: This is for `object` dtype strings (the default), not `string` dtype, so those are still using np.nan (or None)

File: pandas/core/strings/base.py (hunk 8 above)
Comment (TomAugspurger) [reply pandas/core/strings/base.py:13]: Yeah. Probably not worth re-running CI for though. I'll fix it if anything else comes up.

File: pandas/core/arrays/numpy_.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/core/arrays/numpy_.py:407]: And StringArray overrides it to be pd.NA

File: pandas/core/strings/base.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/strings/base.py:13]: sounds good

File: pandas/core/arrays/categorical.py [hunk 9]
```diff
@@ -2312,6 +2313,22 @@ def replace(self, to_replace, value, inplace: bool = False):
         if not inplace:
             return cat
 
+    # ------------------------------------------------------------------------
+    # String methods interface
+    def _str_map(self, f, na_value=np.nan, dtype=np.dtype(object)):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py:2318]: why do these need to be here?

File: pandas/core/strings/accessor.py [hunk 10]
```diff
@@ -0,0 +1,3078 @@
...
+        m = regex.search(x)
+        if m:
+            return [np.nan if item is None else item for item in m.groups()]
+        else:
+            return empty_row
+
+    return f
+
+
+def _result_dtype(arr):
```
Comment (jreback) [inline pandas/core/strings/accessor.py:2945]: why is this here, rather than in the mixins themseleves? e.g. they know there type

File: pandas/core/strings/accessor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/strings/accessor.py:2964]: ok i understand and that is fine. but for the purposes of this PR it is very confusing to put essentially an implementation in the accessor.py, rather than in the concrete classes themselves. alt put them in a separete module for now.

**Review by jreback — CHANGES_REQUESTED** [review]
can you explain how StringMethods *knows* about the object_array and other implementations.

File: pandas/core/arrays/categorical.py (hunk 9 above)
Comment (jreback) [reply pandas/core/arrays/categorical.py:2318]: or rather if you explain / show *how* these are called on the appropriate object

**Comment by jreback** [comment]
> Updated to use a mixin (`ObjectStringArrayMixin`) that implements the pandas string methods on object-dtype ndarrays. EAs implementing the string methods (StringArray, Categorical, and PandasArray for object-dtype strings) inherit from that.

are there concrete mixin's for each of these (even they are empty). I think the structure is not very clear at all.

Please somewhere can you put a diagram of the inheriteance pattern. someone who sees this will be totally confused by the indirection.

File: pandas/core/arrays/categorical.py (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/categorical.py:2318]: @TomAugspurger I would maybe add a comment / docstring here to explain this is overriding the method of `ObjectStringArrayMixin` to ensure the string functions are only applied on the categories and not on the densified values

File: pandas/core/strings/accessor.py (hunk 10 above)
Comment (jorisvandenbossche) [reply pandas/core/strings/accessor.py:2945]: It is here because it is only used in this file .. 
(but that depends on where the implementation for `extract` ends up)

File: pandas/core/strings/accessor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/strings/accessor.py:2964]: > it is very confusing to put essentially an implementation in the accessor.py

I don't think this is confusing, as `extract` is a method of the accessor defined above in this file .. Note that before this PR *all* implementations were in the file where the accessor was defined.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/strings/accessor.py:2964]: Right.

The issue with `extract` specifically is that it operates on the Series rather than the Array. The dispatched functions are all dispatching to the array. I'm sure that we can eventually dispatch to the array by extracting the relevant pieces, but that would be a bit larger change. I've tried to keep the changes here as small as possible (despite the size of the diff) by just moving things and adding just the necessary new pieces.

File: pandas/core/strings/accessor.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/strings/accessor.py:2945]: This is only used in `str_extract`, which isn't currently dispatched (it exists on master as well).

**Comment by TomAugspurger** [comment]
> Please somewhere can you put a diagram of the inheriteance pattern

Added a note to `pandas/core/strings/__init__.py` for the overall layout. The tldr is `Series.str.<method>` calls `Series.array._str_<method>`.

The inheritence is

```python
# BaseStringArrayMethods  # the base ABC
#  - ObjectStringArrayMixin  # most of the implementation
#     - StringArray
#     - PandasArray
#     - Categorical
```

File: pandas/core/strings/object_array.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+                    stacklevel=3,
+                )
+
+            f = lambda x: regex.search(x) is not None
+        else:
+            if case:
+                f = lambda x: pat in x
+            else:
+                upper_pat = pat.upper()
+                f = lambda x: upper_pat in x.upper()
```
Comment (ivanovmg) [inline pandas/core/strings/object_array.py:126]: Is this local variable ``upper_pat`` necessary?

**Comment by jreback** [comment]
> > Please somewhere can you put a diagram of the inheriteance pattern
> 
> Added a note to `pandas/core/strings/__init__.py` for the overall layout. The tldr is `Series.str.<method>` calls `Series.array._str_<method>`.
> 
> The inheritence is
> 
> ```python
> # BaseStringArrayMethods  # the base ABC
> #  - ObjectStringArrayMixin  # most of the implementation
> #     - StringArray
> #     - PandasArray
> #     - Categorical
> ```

great.

let me re-look again, but conceptually looked good.

File: pandas/core/arrays/categorical.py (hunk 9 above)
Comment (jreback) [reply pandas/core/arrays/categorical.py:2318]: yep thanks

Comment (jreback) [reply pandas/core/arrays/categorical.py:2318]: the reason i am asking for this is that our accessor dispatch is already a bit opaque and this is going once step further

File: pandas/core/strings/object_array.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+                # not a compiled regex
+                # set default case
+                if case is None:
+                    case = True
+
+                # add case flag, if provided
+                if case is False:
+                    flags |= re.IGNORECASE
+            if is_compiled_re or len(pat) > 1 or flags or callable(repl):
+                n = n if n >= 0 else 0
```
Comment (ivanovmg) [inline pandas/core/strings/object_array.py:159]: Probably ``n = max(n, 0)``, but maybe not..

File: pandas/core/strings/base.py [hunk 13]
```diff
@@ -0,0 +1,225 @@
+import abc
+from typing import Pattern, Union
+
+import numpy as np
+
+from pandas._typing import Scalar
+
+
+class BaseStringArrayMethods(abc.ABC):
```
Comment (ivanovmg) [inline pandas/core/strings/base.py:9]: Maybe necessary to add abstractmethod ``_str_map``.

Comment (ivanovmg) [reply pandas/core/strings/base.py:9]: Is is necessary to keep both ``BaseStringArrayMethods`` and ``ObjectStringArrayMixin`` if the only class inherited from it is ``ObjectStringArrayMixin``. I  would suggest to move implementation from the child into the parent class and add abstract method ``_str_map``. Unless there is a clear vision on the future extension of ``BaseStringArrayMethods``.

Comment (TomAugspurger) [reply pandas/core/strings/base.py:9]: Our Arrow-backed StringArray (which is motivating this PR) will have separate implementations for most methods. It will inherit directly from BaseStringArrayMethods.

**Comment by TomAugspurger** [comment]
@ivanovmg your suggestions to change some of the implementations are fine, but probably not appropriate for this PR. Right now I'm trying to limit the changes to what's necessary for allowing EAs to controls the implementation, so not changing any of the implementations. Those would be good as followup work.

File: pandas/core/arrays/categorical.py [hunk 14]
```diff
@@ -2316,6 +2316,9 @@ def replace(self, to_replace, value, inplace: bool = False):
     # ------------------------------------------------------------------------
     # String methods interface
     def _str_map(self, f, na_value=np.nan, dtype=np.dtype(object)):
+        # Optimization to apply the callable `f` to the categories once
+        # and rebuild the result by `take`ing from the result with the codes.
+        # Returns the same type as the object-dtype impelmentation though.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/categorical.py:2321]:
```suggestion
        # Returns the same type as the object-dtype implementation though.
```

File: pandas/core/strings/__init__.py [hunk 15]
```diff
@@ -1,3 +1,31 @@
+"""
+Implementation of pandas.Series.str and its interface.
+
+* strings.accessor.StringMethods : Accessor for Series.str
+* strings.base.BaseStringArrayMethods: Mixin ABC for EAs to implement str methods
+
+Most methods on the StringMethods accessor follow the pattern:
+
+    1. extract the array from the series (or index)
+    2. Call that array's impelmentation of the string method
```
Comment (jorisvandenbossche) [inline pandas/core/strings/__init__.py:10]:
```suggestion
    2. Call that array's implementation of the string method
```

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@TomAugspurger needs a merge of master

Otherwise remaining comments?

**Comment by TomAugspurger** [comment]
Fixed the merge conflicts. This should be good on my end.

**Review by simonjayhawkins — COMMENTED** [review]
can pandas/core/strings.py be deleted?

**Comment by TomAugspurger** [comment]
> can pandas/core/strings.py be deleted?

Yeah, apparently it was restored while I was fixing merge conflicts.

File: pandas/core/arrays/categorical.py [hunk 16]
```diff
@@ -2305,6 +2306,25 @@ def replace(self, to_replace, value, inplace: bool = False):
...
+        # Returns the same type as the object-dtype implementation though.
+        from pandas.core.arrays import PandasArray
+
+        categories = self.categories
+        codes = self.codes
+        result = PandasArray(categories.to_numpy())._str_map(f, na_value, dtype)
+        return take_1d(result, codes, fill_value=na_value)
+
+    def _str_get_dummies(self, sep="|"):
+        # sep may not be in categories. Just bail on this.
```
Comment (jbrockmendel) [inline pandas/core/arrays/categorical.py:2323]: does this mean "we are assuming sep is not in self.categories", or "sep might not be in self.categories"?

File: pandas/core/arrays/string_.py [hunk 17]
```diff
@@ -96,7 +103,7 @@ def __from_arrow__(
         return StringArray._concat_same_type(results)
 
 
-class StringArray(PandasArray):
+class StringArray(PandasArray, ObjectStringArrayMixin):
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_.py:106]: ObjectStringArrayMixin is already mixed in to PandasArray, is it needed again here?

File: pandas/core/strings/base.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+    @abc.abstractmethod
+    def _str_fullmatch(
+        self,
+        pat: Union[str, Pattern],
+        case: bool = True,
+        flags: int = 0,
+        na: Scalar = np.nan,
+    ):
+        pass
```
Comment (jbrockmendel) [inline pandas/core/strings/base.py:77]: could `# pragma: no cover` all of these

File: pandas/core/arrays/categorical.py (hunk 16 above)
Comment (TomAugspurger) [reply pandas/core/arrays/categorical.py:2323]: sep might not be in self.categories.

**Comment by TomAugspurger** [comment]
Just the ARM timeout. Anything else here?

File: pandas/core/strings/__init__.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    2. Call that array's implementation of the string method
+    3. Wrap the result (in a Series, index, or DataFrame)
+
+Pandas extension arrays implementing string methods should inherit from
+pandas.core.strings.base.BaseStringArrayMethods. This is an ABC defining
+the various string methods. To avoid namespace clashes and pollution,
+these are prefixed with `_str_`. So ``Series.str.upper()`` calls
+``Series.array._str_upper()``. The interface isn't currently public
+to other string extension arrays.
+"""
```
Comment (jreback) [inline pandas/core/strings/__init__.py:19]: much more clear, thanks

File: pandas/core/strings/accessor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/strings/accessor.py:2966]: ok fair enough, can you create an issue to separate these other methods (maybe with check boxes) & align witht he strutcure)

File: pandas/tests/test_strings.py [hunk 20]
```diff
@@ -728,10 +728,6 @@ def test_count(self):
             ["foo", "foofoo", np.nan, "foooofooofommmfoo"], dtype=np.object_
         )
 
-        result = strings.str_count(values, "f[o]+")
```
Comment (jreback) [inline pandas/tests/test_strings.py:731]: can you also create an issue to either make pandas/tests/strings/test_object.py or add a fixture to support th emultiple implementations (maybe better)

**Review by jreback — APPROVED** [review]
lgtm. some comments to create issues for followon

**Comment by jreback** [comment]
thanks @TomAugspurger really nice. sorry for the long review time :->

**Comment by TomAugspurger** [comment]
Thanks, no worries. This wasn't especially easy to review since it was a
large diff that didn't just move stuff around.

**Comment by xhochy** [comment]
FYI, I used this PR in `fletcher` https://github.com/xhochy/fletcher/pull/196 and it works quite nicely. Will now use that to add string algorithms that are now in Arrow bit-by-bit.

**Comment by TomAugspurger** [comment]
@xhochy Cool! If fletcher is using it then we should probably add it to the public API. Probably just importing it in `pandas.api.extensions`.

**Comment by jorisvandenbossche** [comment]
Unless we only want fletcher to use it for now .. (which still keeps it easier to change the implementation, given the good ties with the fletcher developers?)

**Comment by xhochy** [comment]
I'm going to use it a bit more and see if I run into troubles while overloading it with the Arrow methods. If there are issues, we can easier to breaking changes as long as it is a private API.
